Pr/refactor mips lwip demo

neiljay opened the pull request:
This is a big re-factor of the MIPS lwip based demo (FreeRTOS/Demo/lwIP_MIPS_GCC) moving all common code to a shared common directory.

There are a few changes to generic code and the basic demo as well to make it easier to build for different platforms and ISA revisions.

Further work is required to the coding style to get it to meet the FreeRTOS standard.

Commit:
mips: Use more portable check for assembler

Commit:
mips: demo: Allow overiding of toolchain vars

Allow CROSS_COMPILE and MIPS_ELF_ROOT to be overriden from the environment.

Commit:
mips: use 'addu' not 'add'

addu is more portable and doesn't generate an overflow exception.

Commit:
mips: relax assembler code to make it more portable

Commit:
mips: Add missing declaration

Commit:
mips: uxInterruptNesting is currently unused.

uxInterruptNesting appears to be leftover from the PIC32 port, currently it is
unused in the generic mips port as we are not supporting an interrupt stack yet

Set it to 0 for now as we can hit an assert as this doesn't get decremented on
task start anymore like the comment suggests, but leave it around for when we do
support interrupt stacks.

Commit:
mips: function isn't file static

Commit:
mips: preserve status on context switch.

The _gpctx_load restore code just wholesale copies the
status register from the context back to the register loosing
any changes that may have occured, 'status' is really global state
You dont enable interrupts on one thread and not another...
So we just copy the current status value into the saved value
so nothing changes on the restore.

Commit:
mips: Do not implement _mips_interrupt in C code.

_mips_interrupt is a toolchain provided weak function which gets called on an
interrupt if nothing else has been registered / setup. But implementing it as a C
function is very dangerous as the toolchain provides no context save/restore or
interrupt stack switching before making this call.

For now just let unexpected interrupts go unnoticed.

Commit:
mips: correctly initialise status of new task

When creating a new task use the current status register value for the new task
as interrupt masks are global state.

Commit:
mips: Fix usage of configTICK_RATE_HZ

TICK_RATE_HZ is the frequency of the timer interrupt (eg 1KHz or 100Hz) not the
frequency of the timer/counter used to generate the interrupt.

We have added a new config option 'configTIMER_RATE_HZ' for the freqeunce of the
MIPS count/compare timer (which is typically core clock / 2).

Commit:
mips lwip demo: Big refactor

Create a new common dir, for code common to each platform.
Move uart code to common and rename to represent hardware.
Implement a more generic interrupt handler for VI based systems.
Update python loadscripts (for use with Codescape).

lisaparratt commented on FreeRTOS/Demo/MIPS32_GCC/Makefile:
```diff
@@ -61,22 +61,19 @@ CFLAGS +=-DENABLE_TRACE
 MIPS32_SRC_DIR=../../Source/portable/GCC/MIPS32
-MIPS_ELF_ROOT := $(shell dirname `locate -n 1 $(GCC)`)
-PATH=/usr/bin:/bin:$(MIPS_ELF_ROOT)
+MIPS_ELF_ROOT ?= $(shell dirname `locate -n 1 $(GCC)`)
```
Needs error handling - locate is only common, not omnipresent.

neiljay replied:
Agreed, I don't like this either, the README tells the user to setup MISP_ELF_ROOT anyway so I'll just use the default linux install path (allowing overriding).

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/CI40/ci40_net.c:
```diff
@@ -300,11 +163,106 @@ void ci40_gmac_dma_operation_mode(u32 ioaddr, s32 txmode, s32 rxmode)
+	if (len > HEADER_LEN) {
+		memcpy(eth_tx_buff[current_tx_descriptor]+HEADER_LEN,(void*)&data[HEADER_LEN],len - HEADER_LEN);
+	}
+	mips_flush_dcache();
```
mips_clean_dcache()?

neiljay replied:
Agreed.

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/CI40/ci40_net.c:
```diff
@@ -300,11 +163,106 @@ void ci40_gmac_dma_operation_mode(u32 ioaddr, s32 txmode, s32 rxmode)
+void ci40_net_send(int8_t *data, int32_t len)
+{
+	/* wait for a space */
+	if (dma_tx_desc[current_tx_descriptor].des0.own) {
```
Doesn't the cache need flushing here so .own is up to date?

neiljay replied:
I would say so.

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/CI40/ci40_net.c:
```diff
@@ -300,11 +163,106 @@ void ci40_gmac_dma_operation_mode(u32 ioaddr, s32 txmode, s32 rxmode)
+	dma_tx_desc[current_tx_descriptor].des1.first_segment = 1;
+	dma_tx_desc[current_tx_descriptor].des1.last_segment = 1;
+	dma_tx_desc[current_tx_descriptor].des0.own = 1;
+	mips_flush_dcache();
```
mips_clean_dcache()?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/CI40/ci40_net.c:
```diff
@@ -300,11 +163,106 @@ void ci40_gmac_dma_operation_mode(u32 ioaddr, s32 txmode, s32 rxmode)
+
+	for (i = 0; i < NUM_DMA_DESCS; i ++) {
+		// If filled
+		mips_flush_dcache();
```
mips_clean_dcache()?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/CI40/ci40_net.c:
```diff
@@ -300,11 +163,106 @@ void ci40_gmac_dma_operation_mode(u32 ioaddr, s32 txmode, s32 rxmode)
+	int rx = current_rx_descriptor;
+
+	// If filled
+	if (dma_rx_desc[rx].des0.own == 0) {
```
Doesn't the cache need flushing here?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/common/cpu.h:
```diff
@@ -1,38 +1,40 @@
+
+/* ------------------------ Defines --------------------------------------- */
+#ifndef BYTE_ORDER
+#define BYTE_ORDER LITTLE_ENDIAN
```
Shouldn't this be set based on compiler built-in macros?

neiljay replied:
yes good idea.

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/common/include/arch/cpu.h:
```diff
@@ -1,38 +1,38 @@
+#define __CPU_H__
+
+/* ------------------------ Defines --------------------------------------- */
+#define BYTE_ORDER LITTLE_ENDIAN
```
Shouldn't this be set based on compiler built-in macros?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/common/vi_intc.c:
```diff
@@ -0,0 +1,101 @@
+
+	/* connect up */
+	cpu_irqs[irq->int_num] = irq;
+	pvPortInstallISR( irq->int_num, irq_lookup[irq->int_num & 0x7]);
```
Newer versions of the hal have _mips_intpatch in mips/intctrl.h, which I suspect performs this operation?

neiljay replied:
We need to wait until the latest toolchain release for this, but one to remember.

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/common/vi_intc.c:
```diff
@@ -0,0 +1,101 @@
+
+static HANDLER_DESC_T *cpu_irqs[8];
+
+void intc_isr1(void)
```
_mips_isr_sw0, _mips_isr_sw1, _mips_isr_hw0, _mips_isr_hw1 etc. are weak definitions in the toolchain - if you override them, the core vectors should jump to them. This could be used to avoid manual patching.

neiljay replied:
I'll take a look

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/MicroBlaze-Ethernet-Lite/sys_arch.c:
```diff
@@ -1,584 +1,584 @@
+/* Very crude mechanism used to determine if the critical section handling
+functions are being called from an interrupt context or not.  This relies on
+the interrupt handler setting this variable manually. */
+portBASE_TYPE xInsideISR = pdFALSE;
```
Can't this be done better with SR.IE?

neiljay replied:
All the 'lwip' code is from the lwip release, its in this PR as i dos2unix'd the whole lwIP_MIPS_GCC. I'll remove any modifications to the main lwip source in the next version

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/MicroBlaze-Ethernet-Lite/sys_arch.c:
```diff
@@ -1,584 +1,584 @@
+	}
+	else
+	{
+		while( xSemaphoreTake( *pxSemaphore, portMAX_DELAY ) != pdTRUE );
```
Call _mips_wait() inside loop to reduce power consumption?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/MicroBlaze-Ethernet-Lite/sys_arch.c:
```diff
@@ -1,584 +1,584 @@
+ * @param mutex the mutex to lock */
+void sys_mutex_lock( sys_mutex_t *pxMutex )
+{
+	while( xSemaphoreTake( *pxMutex, portMAX_DELAY ) != pdPASS );
```
Call _mips_wait() inside loop to reduce power consumption?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/WinPCap/Packet32.h:
```diff
@@ -1,359 +1,359 @@
-#endif 
-
-#endif //__PACKET32
+/*
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/WinPCap/PacketData.h:
```diff
@@ -1,267 +1,267 @@
-//	{ pkt15, sizeof( pkt15 ) },
-//	{ pkt16, sizeof( pkt16 ) },
-};
+char pkt1[] = {
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/WinPCap/Win32-Extensions.h:
```diff
@@ -1,113 +1,113 @@
-#endif
-
-#endif //__WIN32_EXTENSIONS_H__
+/*
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/WinPCap/arch.c:
```diff
@@ -1,378 +1,378 @@
-	}
-}
-
+/*
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/WinPCap/netif.h:
```diff
@@ -1,94 +1,94 @@
-portBASE_TYPE xNetifInit( void );
-
-#endif /* NET_IF_H */
+/*
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/WinPCap/remote-ext.h:
```diff
@@ -1,444 +1,444 @@
-
-#endif
-
+/*
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/ethernetif.c:
```diff
@@ -1,657 +1,657 @@
-	xTaskCreate( prvInterruptSimulator, "MAC_ISR", configMINIMAL_STACK_SIZE, NULL, configMAC_ISR_SIMULATOR_PRIORITY, NULL );
-}
-
+/*
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/include/arch/sys_arch.h:
```diff
@@ -1,58 +1,58 @@
-
-#endif /* __ARCH_SYS_ARCH_H__ */
-
+/*
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/lwip/ports/win32/sys_arch.c:
```diff
@@ -1,541 +1,541 @@
- * End of File:  sys_arch.c
- *-------------------------------------------------------------------------*/
-
+/*
```
Superfluous file?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/mipsFPGA/mipsFPGA_eth.c:
```diff
@@ -1,364 +1,312 @@
-
-	//initiate transfer();
-	while (eth->regs[TX_PING_CNTL] & TX_BUSY)
-		;
```
_mips_wait()?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/mipsFPGA/mipsFPGA_eth.c:
```diff
@@ -1,364 +1,312 @@
-	eth->regs[TX_PING_LEN] = len;
-
-	//signal that packet should be sent();
-	eth->regs[TX_PING_CNTL] |= TX_ENABLE; 
```
Cache hasn't been flushed - no guarantee the whole packet is in place?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/mipsFPGA/mipsFPGA_eth.c:
```diff
@@ -1,364 +1,312 @@
-
-	/* Obtain the size of the packet and put it into the "len"
-     variable. */
-	if (eth->regs[RX_PING_CNTL] & RX_READY) {
```
Cache hasn't been flushed?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/mipsFPGA/mipsFPGA_eth.c:
```diff
@@ -1,364 +1,312 @@
-			ptr += q->len;
-		}
-		//acknowledge that packet has been read();
-		eth->regs[RX_PING_CNTL] &= ~RX_READY;
```
Cache hasn't been flushed?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/mipsFPGA/mipsFPGA_eth.c:
```diff
@@ -1,364 +1,312 @@
+	for (;;) {
+		xSemaphoreTake( eth->rx_Semaphore, 0 );
+		mipsFPGA_eth_input(netif);
+		eth->regs[RX_PING_CNTL] &= ~RX_READY;
```
This seems to be done twice?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/mipsFPGA/mipsFPGA_eth.c:
```diff
@@ -1,364 +1,312 @@
+		lwip_stats.link.recv++;
+#endif /* LINK_STATS */
+	} else {
+		/* drop packet(); */
```
Haven't actually dropped the packet in hardware?

lisaparratt commented on FreeRTOS/Demo/lwIP_MIPS_GCC/mipsFPGA/mipsFPGA_intc.c:
```diff
@@ -1,97 +1,105 @@
+
+	regs[AIX_INTC_IAR] =  0x1F;
+	regs[AIX_INTC_IER] =  0x00;
+	pvPortInstallISR( int_num, vIntcISR);
```
_mips_intpatch()?

lisaparratt reviewed (commented):
Mostly looks OK, other than the bits where cache flushing is missed out completely. Most of the other comments are just food for thought.

Commit:
mips: Tidy up timer after merge.

Note we have gone back to the original timer implementation.
For the basic demo we allow all 3 clock parameters to be configurable from the
.config file (core clock, timer clock and tick frequency), typically the
count/comapre timer runs at half core clock but it is allowed to be different.

Commit:
mips: mipsFPGA ethernet: don't ack the packet twice.

Commit:
mips: Ci40_net: Clean up cache flush code.

Only flush the required data not the whole cache and add missing invalidate's.

Commit:
gitignore: ignore some more stuff: .py .swp .a

Commit:
mips: lwip demo: dos2unix
